xgcm/sgrid.py

```diff
@@ -0,0 +1,230 @@
+        (x for x in ["Conventions", "conventions"] if x in ds.attrs), False
+    )
+    if conventions_attr:
+        if any(x in ds.attrs[conventions_attr] for x in ["SGRID", "sgrid", "Sgrid"]):
+            return True
+
+    return False
+
+
+def get_sgrid_grid(ds):
+    """Extract the name of the variable containing the SGRID grid data.
+    This can be done by checking cf_role attribute for 'grid_topology'
+
+    Parameters
+    ----------
+    ds : xarray.dataset
+
+    Returns
+    ----------
+    var_name : str
+        The name of the grid variable
+    """
+
+    for var_name in ds.data_vars:
+        if ds[var_name].attrs.get("cf_role") == "grid_topology":
+            return var_name
+    raise ValueError("Could not find identify SGRID grid in input dataset.")
+
+
+def get_all_axes(ds):
+    """Works out how many axes there are and what the names of them are.
+    Names are added to a set.
+
+    Parameters
+    ----------
+    ds : xarray.dataset
+
+    Returns
+    ----------
+    axes : set
+        set of strings containing the names of the spatial axes in the dataset
+
+    """
+    axes = set()
+
+    sgrid_grid_name = get_sgrid_grid(ds)
+    ndims = ds[sgrid_grid_name].attrs["topology_dimension"]
+    if ndims == 1:
+        axes.update(["X"])
+    elif ndims == 2:
+        axes.update(["X", "Y"])
+        # Check for a vertical dimension
+        if "vertical_dimensions" in ds[sgrid_grid_name].attrs:
+            axes.update(["Z"])
+    elif ndims == 3:
+        axes.update(["X", "Y", "Z"])
+    else:
+        raise ValueError(
+            f"SGRID dimensions {ndims} in variable {sgrid_grid_name} is > 3."
+        )
+    return axes
+
+
+def get_axis_positions_and_coords(ds, axis_name):
+    """SGRID version of related function in `comodo.py`.
+    Returns an ordered dictionary of position->coordinate dimension entries
+    associated with a given axis, which is then stored in the `Axis.coords`
+    instance variable.
+    Populate with keywords 'center', 'inner', 'outer', 'left', 'right'
+    corresponding to the name of the dimensions on this axis.
+    For SGRID this can be done based on whether it is node or face and looking
+    at the padding attribute.
+
+    1) Node dimensions are guaranteed by basic SGRID so extract these first
+    2) Face or Volume dimensions are guaranteed and are associated with the cell
+       coordinates (center)
+        a) Check grid dimension and extract face/vol appropriately
+        b) Syntax match face/vol dimensions to node dimension
+        c) Record Shift and assign to nodes as appropriate
+    3) Facei and edgei are optional and depend on 2D vs 3D.
+       They add no additional information about dimensions so ignored for now.
+
+    SGRID cell padding types can be converted to xgcm node positions as follows:
+      - padding low  refers to 'right'
+      - padding high refers to 'left'
+      - padding both refers to 'inner'
+      - padding none refers to 'outer'
```

Comment:
This is a very useful Rosetta Stone to understand relationship between SGRID and xgcm. Thanks for writing it up! :+1: